// ms-msal-angular/dist/msal.interceptor.js
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/fromPromise';
import 'rxjs/add/operator/mergeMap';
import { MsalService } from "./msal.service";
import { BroadcastService } from "./broadcast.service";
export class MsalInterceptor {
    constructor(auth, broadcastService) {
        this.auth = auth;
        this.broadcastService = broadcastService;
    }
    intercept(req, next) {
        var scopes = this.auth.get_scopes_for_endpoint(req.url);
        this.auth.verbose('Url: ' + req.url + ' maps to scopes: ' + scopes);
        if (scopes === null) {
            return next.handle(req);
        }
        var tokenStored = this.auth.getCached_Token_Internal(scopes);
        if (tokenStored && tokenStored.token) {
            req = req.clone({
                setHeaders: {
                    Authorization: `Bearer ${tokenStored.token}`,
                }
            });
            return next.handle(req);
        }
        else {
            return Observable.fromPromise(this.auth.acquire_token_silent(scopes).then(token => {
                const JWT = `Bearer ${token}`;
                return req.clone({
                    setHeaders: {
                        Authorization: JWT,
                    },
                });
            })).mergeMap(req => next.handle(req)); //calling next.handle means we are passing control to next interceptor in chain
        }
    }
}
MsalInterceptor.decorators = [
    { type: Injectable },
];
/** @nocollapse */
MsalInterceptor.ctorParameters = () => [
    { type: MsalService, },
    { type: BroadcastService, },
];
